carp.data.core/src/commonMain/kotlin/dk/cachet/carp/data/application/DataStreamBatch.kt

```diff
@@ -134,6 +141,149 @@ class MutableDataStreamBatch : DataStreamBatch
+    {
+        /**
+         * Create an [ImmutableDataStreamBatch] from the provided sequences.
+         *
+         * @throws IllegalArgumentException when sequences within the same DataStreamId overlap.
+         *
+         * note: Sequences across different devices/data types may overlap (this is expected for analytics).
+         * note: private constructor + factory methods provides validation and optimal internal structure.
+         */
+        fun create( sequences: Sequence<DataStreamSequence<*>> ): ImmutableDataStreamBatch
+        {
+            // Validate that sequences within the same DataStreamId don't overlap
+            val sequencesByDataStream = sequences.groupBy { it.dataStream }
+            sequencesByDataStream.forEach { (dataStreamId, streamSequences) ->
+                val sortedSequences = streamSequences.sortedBy { it.range.first }
+                for (i in 1 until sortedSequences.size)
+                {
+                    val previous = sortedSequences[i - 1]
+                    val current = sortedSequences[i]
+                    require(previous.range.last < current.range.first) {
+                        "Sequences within data stream $dataStreamId overlap: ${previous.range} and ${current.range}"
+                    }
+                }
+            }
+
+            // Build hierarchical structure: Device → DataType → Sequences
+            val deviceDataMap = sequences
+                .groupBy { it.dataStream.deviceRoleName }
+                .mapValues { (_, deviceSequences) ->
+                    deviceSequences.groupBy { it.dataStream.dataType }
+                }
+
+            return ImmutableDataStreamBatch(deviceDataMap)
+        }
+
+        /**
+         * Create an [ImmutableDataStreamBatch] from an existing [DataStreamBatch].
+         */
+        fun from( batch: DataStreamBatch ): ImmutableDataStreamBatch =
+            create(batch.sequences)
+
+        /**
+         * Create an empty [ImmutableDataStreamBatch].
+         */
+        fun empty(): ImmutableDataStreamBatch =
+            ImmutableDataStreamBatch(emptyMap())
+    }
+
+    /**
+     * Get all device role names represented in this batch.
+     */
+    fun getDeviceRoleNames(): Set<String> = deviceDataMap.keys
+
+    /**
+     * Get all data types represented in this batch.
+     *
+     * @param deviceRoleName if specified, only return data types for this device role name
+     */
+    fun getDataTypes( deviceRoleName: String? = null ): Set<DataType> =
+        if (deviceRoleName != null)
+        {
+            deviceDataMap[deviceRoleName]?.keys.orEmpty()
+        } else
+        {
+            deviceDataMap.values.flatMap { it.keys }.toSet()
+        }
+
+    /**
+     * Get all sequences for a specific device role name.
+     *
+     * @param synchronize if true, convert all sequences to synchronized data points and flatten
+     */
+    fun getForDevice( deviceRoleName: String, synchronize: Boolean = false ): Any =
```

Comment:
Synchronization is an operation. And, _how_ to synchronize is open for change. Directly linking that to the data structure is thus undesirable. Instead, have an operation perform synchronization _on_ the data structure.

Why is the return type `Any`?